Assets/UI/worldinput_CQUI.lua

```diff
@@ -0,0 +1,559 @@
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.H and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_CAMP"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.I and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FARM"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.L and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_LUMBER_MILL"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.N) then
+    cquiHandledKey = true;
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_MINE"].Hash);
+    else
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.WMD_STRIKE);
+    end
+  end
+
+  if (uiKey == Keys.O) then
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_OIL_WELL"].Hash);
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_OFFSHORE_OIL_RIG"].Hash);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.P) then
+    if (CQUI_isShiftDown) then
+      CQUI_isShiftDown = false;
+      PlaceMapPin();
+      cquiHandledKey = true;
+    else
+      if (unitType == "UNIT_BUILDER") then
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PASTURE"].Hash);
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PLANTATION"].Hash);
+        cquiHandledKey = true;
+      end
+    end
+  end
+
+  if (uiKey == Keys.R) then
+    if (unitType == "UNIT_MILITARY_ENGINEER") then
+      -- Build Road/Rail is a UnitOperation
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_BUILD_ROUTE"].Hash);
+      cquiHandledKey = true;
+    elseif (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REBASE);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.S and CQUI_isAltDown == true) then
+    if (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitOperationTypes.AIR_ATTACK);
+      -- Use false as we need to let the unmodifed script deal with the Alt Key
+      cquiHandledKey =false;
+    end
+  end
+
+  if uiKey == Keys.VK_SHIFT then
+    CQUI_isShiftDown = false;
+  end
+
+  if uiKey == Keys.VK_ALT then
+    CQUI_isAltDown = false;
+  end
+
+  if (cquiHandledKey == false) then
+    print_debug("CQUI Did not handle key: "..tostring(uiKey));
+    cquiHandledKey = BASE_CQUI_DefaultKeyUpHandler( uiKey );
+  end
+
+  return cquiHandledKey;
```

Comment:
Old version would only return true when '( uiKey == Keys.VK_ADD or uiKey == Keys.VK_SUBTRACT )'. Any reason for change? Is this supposed to return true whenever a keystroke is consumed?

Reply:
yep.  when the keystroke is handled, returning true prevents it from going to other handlers (for example, pushing W in Exp2.  When you have a great work, W is the default show Great Works key.  If we do not return true from this handler, then the Great Work panel is opened.  

Which reminds me, adding Alt-W to open the great work panel may be a thing to do

Reply:
I've decided against the Alt-W thing for the time being